### PR TITLE
Facilitate callbacks on error and scenario/feature success

### PR DESCRIPTION
**Pull request by dokie**
Signed-off-by: Dr. Mike Williams <m.williams@suez.com>

### CODE REVIEW

File: apps/ex_cucumber/lib/ex_cucumber.ex [hunk 1]
```diff
@@ -188,9 +198,22 @@ defmodule ExCucumber do
...
           end
         end
       end
+
+      defp callback_on_error(context) do
+        if Module.has_attribute?(__MODULE__, :on_error) do
+          on_error = Module.get_attribute(__MODULE__, :on_error)
+          if is_function(on_error, 1), do: on_error.(context)
+        end
+      end
```
Comment (slashdotdash) [inline apps/ex_cucumber/lib/ex_cucumber.ex]: Do we also want to put a `try` ... `rescue` / `catch`  block around the user-supplied callback functions to ensure we also catch errors raised there and handle them appropriately?

**Review by slashdotdash — APPROVED** [review]

File: apps/ex_cucumber/lib/ex_cucumber.ex (hunk 1 above)
Comment (slashdotdash) [reply apps/ex_cucumber/lib/ex_cucumber.ex]: We might also want to add a warning if the module attribute is not a one-arity function.

```elixir
defp callback_on_error(context) do
  case Module.get_attribute(__MODULE__, :on_error) do
    nil -> 
      :ok

    on_error when is_function(on_error, 1) -> 
      on_error.(context)

    _invalid -> 
      # Raise or log error
      raise ArgumentError, message: "invalid `@on_error` module attribute in: ` <> inspect(__MODULE__)
  end
end
```

File: apps/ex_cucumber/lib/ex_cucumber/gherkin/traverser/feature_traverser.ex [hunk 2]
```diff
@@ -6,27 +6,77 @@ defmodule ExCucumber.Gherkin.Traverser.Feature do
...
         Map.merge(feature_meta(f), %{state: %{}, step_history: [], context_history: []})
       )
 
-    children
-    |> Enum.each(fn child ->
-      acc = MainTraverser.run(background, acc, parse_tree)
+    result =
+      children
+      |> Enum.reduce(merged_ctx, fn child, ctx ->
+        uodated_ctx = MainTraverser.run(background, ctx, parse_tree)
```
Comment (gesta) [inline apps/ex_cucumber/lib/ex_cucumber/gherkin/traverser/feature_traverser.ex]:
```suggestion
        updated_ctx = MainTraverser.run(background, ctx, parse_tree)
```
and several other places.

File: apps/ex_cucumber/test/support/modules/rule_feature/demonstrate_success_callbacks.ex [hunk 3]
```diff
@@ -0,0 +1,59 @@
+defmodule Support.RuleFeature.DemonstrateSuccessCallbackse do
```
Comment (gesta) [inline apps/ex_cucumber/test/support/modules/rule_feature/demonstrate_success_callbacks.ex]:
```suggestion
defmodule Support.RuleFeature.DemonstrateSuccessCallbacks do
```

**Review by gesta — APPROVED** [review]
Couple of typos to address, but looks good. 🏁

**Review by gesta — APPROVED** [review]

**Pull request merged**